Replace automatic bans with discouragement filter

sipa opened the pull request:
This patch improves performance and resource usage around IP addresses that are banned for misbehavior. They're already not actually banned since #14929, as connections from them are still allowed, but they are preferred for eviction if the inbound connection slots are full.

Stop treating these like manually banned IP ranges, and instead just keep them in a rolling Bloom filter of misbehaving nodes, which isn't persisted to disk or exposed through the ban framework. The effect remains the same: preferred for eviction, avoided for outgoing connections, and not relayed to other peers.

Also change the name of this mechanism to "discouraged" to better reflect reality.

sipa committed:
Replace automatic bans with discouragement filter

This patch improves performance and resource usage around IP
addresses that are banned for misbehavior. They're already not
actually banned, as connections from them are still allowed,
but they are preferred for eviction if the inbound connection
slots are full.

Stop treating these like manually banned IP ranges, and instead
just keep them in a rolling Bloom filter of misbehaving nodes,
which isn't persisted to disk or exposed through the ban
framework. The effect remains the same: preferred for eviction,
avoided for outgoing connections, and not relayed to other peers.

Also change the name of this mechanism to better reflect reality;
they're not banned, just discouraged.

Contains release notes and several interface improvements by
John Newbery.

luke-jr reviewed (changes requested):
Concept ACK

I think we need to skip over BanReasonNodeMisbehaving entries when loading existing ban data?

jonatack commented on src/banman.h:
```diff
@@ -68,6 +77,7 @@ class BanMan
     CClientUIInterface* m_client_interface = nullptr;
     CBanDB m_ban_db;
     const int64_t m_default_ban_time;
+    CRollingBloomFilter m_discouraged{50000, 0.000001};
```
A comment on the choice of nElements and nFPRate may be good here. These are the same as for `filterInventoryKnown`.

jonatack reviewed (commented):
Concept ACK on improving the performance/resource usage of IsBanned.

Light code review ACK 4c4df6bf61104a4946479fccc7885dc0fe05dbd8. I have not benchmarked the performance deltas of IsBanned (presumably improved for discouraged addrs) and IsBannedLevel (presumably marginally worse for non-banned addrs) and how meaningful they are with this change.

MarcoFalke commented on src/test/denialofservice_tests.cpp:
```diff
@@ -346,12 +346,6 @@ BOOST_AUTO_TEST_CASE(DoS_bantime)
     }
     BOOST_CHECK(banman->IsBanned(addr));
 
```
```suggestion
     BOOST_CHECK_EQUAL(1, banman->IsBannedLevel(addr));
```

nit: Could add a test for `IsBannedLevel`?

maflcko reviewed (commented):
review ACK 4c4df6bf61104a4946479fccc7885dc0fe05dbd8 🚆

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 4c4df6bf61104a4946479fccc7885dc0fe05dbd8 🚆
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiFHgv/X2KHSRqmN8x9c5zLH7/3ZqkSkcMGsr1wMBS3GXDqAYv14lYl2xAG4iou
TAkeccq8MLDpARLeTtjVou1NhjFcEMZXciV55nF4y7GD8DianWdAeH0/yTlGfjdk
yuYGZJpQlHipSccFgOY+pf5pTS9Ph1XgHEabfU2mwGsMRHeo5wIGulBGHTH3oGPZ
r8aTkNv2o0hQ7Gm68g5k8+cYCcnPYTy8CdO/3CTJMjzY7fgg9ig/CuOWWAq3Pka1
IcxspupCNtwYKui+rFMXpLar78b1ABTLKGRX2iHBLs09bdOMZgO5jDDCooWx7YAV
ogNVxmb93p7kRiVtZschxRCnGDBnn6TGJKINNxKwfuwu8BPHilHOwWXmiKUF7EKh
MwVsVQGYrGz1+uSst4FYhwEXewC6o72kC/7t58mKy2hipyFhMhjQC/CW8v0M9Xsu
K4CIviPL/AphwRDhTff6nk1uot9VJHaC/JXxaJmnaXiYkckdlu9D2VaKe5n5dm58
gy6VugHp
=eVeU
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `efbc9feace2c4d03c3d87e850e69600a2022027bb9fda48388d15e2d99cfe379  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108efbc9feace2c4d03c3d87e850e69600a2022027bb9fda48388d15e2d99cfe379f01057893df9b5454aa1716abfef79186eb008fff0105cb06e5efb1f6793e82592c1bd8c518608f1045edf8f10f0083f4be2f4cf47abec0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105e6f59a290d265c8994742094de4741508f1045edf8f10f008acd0e943fe95e4be0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f01068422c4f7600171893be94af16e2971608f1045edf8f11f00802929601a685f6070083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

laanwj commented:
Code review ACK 4c4df6bf61104a4946479fccc7885dc0fe05dbd8

sipa commented:
@luke-jr I don't think that's necessary. They (by default) have a 24 hour limit only, and during that time, they will temporarily be treated as a real ban instead of a discouragement.

pstratem commented on src/banman.cpp:
```diff
@@ -72,27 +72,26 @@ int BanMan::IsBannedLevel(CNetAddr net_addr)
         }
     }
-    return level;
+    return m_discouraged.contains(net_addr.GetAddrBytes());
```
it feels a bit odd to treat what is essentially an enum (0/1/2) as an int like this, suggest replacing with a conditional returning 1/0 explicitly

sipa replied:
Ok, added an explicit `? 1 : 0`.

jasonbcox commented on src/banman.h:
```diff
@@ -45,10 +46,18 @@ class BanMan
     void Ban(const CNetAddr& net_addr, const BanReason& ban_reason, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
     void Ban(const CSubNet& sub_net, const BanReason& ban_reason, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
+    void Discourage(const CNetAddr& net_addr);
     void ClearBanned();
```
Looks like ClearBanned() should call `m_discouraged.reset()` otherwise the `clearbanned` RPC no longer does what users expect.

sipa replied:
I consider that, but I think it would be counterintuitive. Discouragement isn't exposed as bans (it's not visible or accessible in RPCs or UI) as it's a completely automatic process, so I think that exposing a way to clear it would violate that abstraction.

jasonbcox replied:
I see. The current implementation may introduce subtle bugs then, especially considering that `IsBanned()` is now unintuitively named. `setban "someIP" "add"` may return `Error: IP/Subnet already banned` for instance.

sipa replied:
Nice catch. I've addressed it by adding a separate `IsBannedOrDiscouraged` that's distinct from `IsBanned`.

jnewbery commented on src/banman.h:
```diff
@@ -68,6 +77,7 @@ class BanMan
     CClientUIInterface* m_client_interface = nullptr;
     CBanDB m_ban_db;
     const int64_t m_default_ban_time;
+    CRollingBloomFilter m_discouraged{50000, 0.000001};
```
Should this have a `GUARDED_BY(m_cs_banned)` lock annotation?

sipa replied:
Done.

jnewbery commented:
This seems like an improvement, but it leaves the `BanMan` with a slightly odd and inconsistent interface:

- `Ban()` has an enum as its second parameter, but that parameter is only ever set to `BanReasonManuallyAdded`
- If `IsBanned()` is called with an address, it returns whether an address is discouraged or banned. If it's called with a subnet, it returns if it's banned, but not discouraged.
- `Unban()` won't remove a discouraged address from banman, but it's not possible to add an outbound connection to that address with `addnode onetry`.
- `ClearBanned()` (which is called by the `clearban` RPC ("Clear all banned IPs")) doesn't clear discouraged addresses.
- The `listbanned` RPC returns a `ban_reason`, which can only ever be "manually added". (`banReasonToString()` is basically dead code now).

It'd be nice to rationalize this so BanMan just had a setter `SetBanLevel()` function and a getter `GetBanLevel()` function.

sipa commented:
@jnewbery Most of those things are no longer true, apart from the BanReason being superfluous now (there is a separate IsBannedOrDiscouraged). I think a follow-up can remove BanReason altogether, but I think it's preferable to not have a SetBanLevel - discouragement and banning are very different concepts.

pstratem commented:
ACK 4131caf2431054adf9ad6641386dfbe922af0dc5

sipa committed:
Clean up separated ban/discourage interface

sipa commented:
I've made a few changes suggested by @jnewbery. The PR is now split into two commits:
* The first is a minimal change that implements the new behavior (smaller than the previous PR), plus release notes and RPC documentation changes.
* The seconds makes the distinction between discouraging and banning much cleaner in the interface, removing BanReason altogether.

I'm happy to split off the second commit. It could also be ignored in backports.

luke-jr commented on doc/release-notes-19219.md:
```diff
@@ -0,0 +1,20 @@
+#### Change in automatic banning
+
+Automatic banning of peers for bad behavior has been slightly altered:
```
Since they're not really bans, let's not call them bans in docs?

sipa replied:
I think this is improved now with @jnewbery's changes.

pstratem commented on src/net.cpp:
```diff
@@ -1010,11 +1010,12 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     // Don't accept connections from banned peers, but if our inbound slots aren't almost full, accept
     // if the only banning reason was an automatic misbehavior ban.
-    if (!NetPermissions::HasFlag(permissionFlags, NetPermissionFlags::PF_NOBAN) && bannedlevel > ((nInbound + 1 < nMaxInbound) ? 1 : 0))
+    if (!NetPermissions::HasFlag(permissionFlags, NetPermissionFlags::PF_NOBAN) && (banned || (discouraged && nInbound + 1 < nMaxInbound)))
```
maybe this should me two checks so the log can indicate ban/discouraged

sipa replied:
This has been rewritten following a suggestion by @pstratem below.

pstratem commented:
ACK 0778faa584871010c82db84c9d54d1353a91a37c

jonatack commented on src/banman.h:
```diff
@@ -68,6 +77,17 @@ class BanMan
     CBanDB m_ban_db;
     const int64_t m_default_ban_time;
+
+    /** Individual addresses that of peers that misbehave (see Misbehaving() in
```
b6a34484 nit: omit "that"

sipa replied:
This is gone.

jonatack commented on src/banman.h:
```diff
@@ -68,6 +77,17 @@ class BanMan
+    /** Individual addresses that of peers that misbehave (see Misbehaving() in
+     *  net_processing) are not banned but discouraged: we still allow incoming
+     *  connections from them, but prefer them for eviction. Similar to banned
+     *  IPs, we never make outgoing connections to them, and do not rumour them
```
b6a34484 For my understanding, "rumour" == gossip?

sipa replied:
This is gone.

jonatack commented on src/rpc/net.cpp:
```diff
@@ -671,7 +672,6 @@ static UniValue listbanned(const JSONRPCRequest& request)
         rec.pushKV("address", entry.first.ToString());
         rec.pushKV("banned_until", banEntry.nBanUntil);
         rec.pushKV("ban_created", banEntry.nCreateTime);
-        rec.pushKV("ban_reason", banEntry.banReasonToString());
```
b6a34484 Should also remove `ban_reason` from the help at line 650, I believe.

sipa replied:
Done.

MarcoFalke replied:
Have the release notes been updated with this breaking RPC change?

sipa replied:
Fixed.

jonatack reviewed (commented):
Reviewing each commit separately.

ACK commit b6a34484 modulo removing `{RPCResult::Type::STR, "ban_reason", ""},` from `src/rpc/net.cpp::L650`

jonatack reviewed (commented):
Nice cleanup in 0778faa584871010c82db84c9d54, a few comments follow.

luke-jr commented on doc/release-notes-19219.md:
```diff
@@ -0,0 +1,20 @@
+- automatic bans will no longer be returned by the `listbanned` RPC.
+
+- automatic bans can no longer be lifted with the `setban remove` RPC command.
+  If you need to remove an automatic ban, you can clear all bans (including
+  manual bans) with the `clearbanned` RPC, or stop-start to clear automatic bans.
```
I think it would be useful to have a way to clear only the misbehaving filter, but perhaps that is best left for a followup PR.

sipa replied:
Agree.

jnewbery commented:
There are quite a few places where comments/names/logs are now out of date. Commit here: https://github.com/jnewbery/bitcoin/tree/pr19219.2 gets most of them I think.

sipa commented on src/net.cpp:
```diff
@@ -1010,11 +1010,12 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
+    bool banned = m_banman && m_banman->IsBanned(addr);
 
     // Don't accept connections from banned peers, but if our inbound slots aren't almost full, accept
     // if the only banning reason was an automatic misbehavior ban.
```
This has been split up, so I don't think this applies anymore.

sipa commented:
@jnewbery I've incorporated most of the changes from your branch (split out over both commits).

I dropped the PF_NOBAN -> PFNODISCOURAGE change, because that capability also protects against bans.

jnewbery commented on src/addrdb.h:
```diff
@@ -44,31 +36,17 @@ class CBanEntry
     {
-        banReason = ban_reason_in;
+        uint8_t ban_reason = 2; //! For backward compatibility
+        READWRITE(obj.nVersion, obj.nCreateTime, obj.nBanUntil, ban_reason);
```
Note for reviewers: this means that if any peers are discouraged at the time that the node was last shut down pre-upgrade, they'll be loaded as banned post-upgrade. They'll clear automatically after 24 hours, so I think this is ok.

jnewbery commented on src/banman.h:
```diff
@@ -23,32 +24,55 @@ class CClientUIInterface;
     BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time);
-    void Ban(const CNetAddr& net_addr, const BanReason& ban_reason, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
-    void Ban(const CSubNet& sub_net, const BanReason& ban_reason, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
+    void Ban(const CNetAddr& net_addr, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
```
It should be very straightforward to unify the `Ban()` public functions and `IsBanned()` public functions so that the caller always passes in a (possibly /32) subnet. That would actually make the calling code in `setban` simpler.

This is out of scope for this PR, but could be done in a follow-up.

jnewbery commented on src/net.cpp:
```diff
@@ -1010,17 +1010,25 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
 
-    int bannedlevel = m_banman ? m_banman->IsBannedLevel(addr) : 0;
+    bool banned = m_banman->IsBanned(addr);
+    bool discouraged = banned || m_banman->IsDiscouraged(addr);
```
or'ing with `banned` is confusing here. After this PR, banning and discouraging are totally separated concepts, but this code signals that discouraging is a superset of banning.

sipa replied:
I was trying to avoid a call to IsDiscouraged if the peer is already banned. I moved it to accomplish something similar. I think it's clearer now.

jnewbery commented on src/net.cpp:
```diff
@@ -2046,7 +2054,7 @@ void CConnman::OpenNetworkConnection(const CAddress& addrConnect, bool fCountFai
     if (!pszDest) {
         if (IsLocal(addrConnect) ||
-            FindNode(static_cast<CNetAddr>(addrConnect)) || (m_banman && m_banman->IsBanned(addrConnect)) ||
+            FindNode(static_cast<CNetAddr>(addrConnect)) || (m_banman && (m_banman->IsDiscouraged(addrConnect) || m_banman->IsBanned(addrConnect))) ||
```
This would be easier to read if `(m_banman .... )` conditional was on its own line, so it's easier to see that the `||` within that conditional is not at the top level.

sipa replied:
Done.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3559,25 +3560,25 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
             pnode.fDisconnect = true;
         } else {
-            // Disconnect and ban all nodes sharing the address
+            // Disconnect and discourage all nodes sharing the address
```
You added logging here in the first commit and then (accidentally, I think) removed it in the second.

sipa replied:
Fixed.

jnewbery commented on src/banman.h:
```diff
@@ -23,32 +24,55 @@ class CClientUIInterface;
+// banlist.dat on shutdown and reloaded on startup. Banning can be used to
+// prevent connections with spy nodes or other griefers.
+//
+// 2. Discouragement. If a peer misbehaves enough (see Misbehaving()) in
```
nit: the parens are in the wrong place - this should say "(see Misbehaving() in net_processing.cpp)

Sorry, this is my fault!

sipa replied:
Fixed.

jnewbery commented on src/rpc/net.cpp:
```diff
@@ -556,7 +556,8 @@ static UniValue getnetworkinfo(const JSONRPCRequest& request)
     const RPCHelpMan help{"setban",
-                "\nAttempts to add or remove an IP/Subnet from the banned list.\n",
+                "\nAttempts to add or remove an IP/Subnet from the banned list.\n"
+                "Peers that are automatically banned cannot be unbanned.\n",
```
This line should be removed (sorry - this was from my branch when I was referring to automatic banning)

sipa replied:
Done.

jnewbery reviewed (commented):
A few small style suggestions, otherwise looks good.

jnewbery commented:
utACK 250e8acc7df63430f73099549d18d707844c08a8

jnewbery commented:
utACK 0ba85217caa333021cf04eb5cff9770f3152c287

jonatack commented on src/net.cpp:
```diff
@@ -1010,17 +1010,24 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
 
+    // Only accept connections from discouraged peers if our inbound slots aren't (almost) full.
+    bool discouraged = m_banman->IsDiscouraged(addr);
+    if (!NetPermissions::HasFlag(permissionFlags, NetPermissionFlags::PF_NOBAN) && nInbound + 1 >= nMaxInbound && discouraged)
```
The conditional for determining if the inbounds are nearly full, per previous review https://github.com/bitcoin/bitcoin/pull/19219#discussion_r438752687, now seems correct.

sipa replied:
Ok.

jonatack reviewed (commented):
ACK 0ba85217caa333021cf04eb5cff9770f3152c287 code review, verified builds, tests, ran updated fuzzer (`src/test/fuzz/addrdb ../qa-assets/fuzz_seed_corpus/`). Am running bitcoind on mainnet with this branch and banscore at low thresholds (1, 0), grepping getpeerinfo banscores (`bitcoin-cli getpeerinfo | grep 'banscore'`) and the net debug log (`grep -ni "bann\|discour" ~/.bitcoin/debug.log`), but the banscores have remained at 0 so far and no discouragement has been logged. As a sanity check did some manual setbans, checked listbanned and the log, watched the bans expire.

Some suggestions if you retouch or as a follow-up.

cculianu commented on src/net.cpp:
```diff
@@ -1044,7 +1051,7 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
     // If this flag is present, the user probably expect that RPC and QT report it as whitelisted (backward compatibility)
     pnode->m_legacyWhitelisted = legacyWhitelisted;
-    pnode->m_prefer_evict = bannedlevel > 0;
+    pnode->m_prefer_evict = banned || discouraged;
```
What if they are `PF_NOBAN`? .. these flags may be set then and thus `m_prefer_evict` will be set... is that acceptable?

sipa replied:
PF_NOBAN peers are already except from eviction entirely, so the value of the `m_prefer_evict` field has no meaning for them.

The code here is just maintaining the existing semantics (current master also sets `m_prefer_evict` for (auto)banned peers with PF_NOBAN).

sipa commented on src/init.cpp:
```diff
@@ -427,8 +427,8 @@ void SetupServerArgs(NodeContext& node)
-    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-bantime=<n>", strprintf("Number of seconds to keep misbehaving peers from reconnecting (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting and discouraging misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-bantime=<n>", strprintf("Default bantime for manually configured bans (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
I don't understand what your comment on the use of "default" is about. It can indeed by overridden by providing an RPC argument.

I've added seconds back.

jonatack replied:
Thanks. The two uses of "default" with different meanings in the same sentence (default for manual bans and default for this setting, e.g. the default for the default) seemed confusing to me. That said, reading it with "(in seconds)" now separating the two seems better.

sipa commented on src/banman.h:
```diff
@@ -23,32 +24,55 @@ class CClientUIInterface;
+// 1. Banning. This is configured manually by the user, through the setban RPC.
+// If an address or subnet is banned, we never accept incoming connections from
+// it and never create outgoing connections to it. We won't gossip its address
+// to other peers in addr messages. Banned addresses and subets are stored to
```
Fixed.

sipa commented on doc/release-notes-19219.md:
```diff
@@ -0,0 +1,20 @@
+eviction. There are some small changes to our treatment of discouraged
+addresses:
+
+- Discouraging an address does not time out automatically after 24 hours.
```
Done.

sipa commented on doc/release-notes-19219.md:
```diff
@@ -0,0 +1,20 @@
+#### Change in banning behaviour
```
Done.

sipa commented on src/init.cpp:
```diff
@@ -427,8 +427,8 @@ void SetupServerArgs(NodeContext& node)
     gArgs.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by the net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-bantime=<n>", strprintf("Number of seconds to keep misbehaving peers from reconnecting (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting and discouraging misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
I think `banscore` should die a swift death (but not in this PR).

jnewbery commented:
utACK 2605bdd220b118fe3cac6a9cebb673cf651e273c

jonatack commented:
Code review re-ACK 2605bdd per git diff `0ba8521 2605bdd` changes since last review are documentation and help improvements and a code change to clarify that only discouragement affects the `bool CNode::m_prefer_eviction` value for peers, as we don't accept connections from banned peers.

narula commented on src/banman.cpp:
```diff
@@ -151,8 +137,8 @@ void BanMan::Ban(const CSubNet& sub_net, const BanReason& ban_reason, int64_t ba
     }
     if (m_client_interface) m_client_interface->BannedListChanged();
 
-    //store banlist to disk immediately if user requested ban
```
nit: I think this change makes more sense in the previous commit, since that's where you added the assert.

sipa replied:
Agreed. Will fix if there are more changes requested, but I don't think this is worth breaking ACKs for.

sipa replied:
Done.

amitiuttarwar reviewed (commented):
#### high level review thoughts 
strong concept / approach ACK. splitting out separate bloom filter from the (now manual) ban list makes sense conceptually and ofc is a robustness improvement. 

over the last week, I have spent _many_ hours reviewing this PR & learning about the surrounding mechanisms. from what I can tell, the functionality looks solid.

even after spending the week with this code, I still find the `discourage` language & logical event flow to be kind of confusing. we compare a misbehavior score to a ban score to figure out if the peer should be discouraged, which usually means "add to discouragement filter", but in this case means "possibly disconnect and possibly add to discouragement filter" 😅

I think part of the problem is the verb "discourage" makes it seem like we are taking an explicit action towards that connection. I think language like "mark as discouraged" or "deprioritize" would better match the meaning of add-to-filter-that-we-later-check. even better would be making it cohesive with the misbehaving mechanisms (since they seem to exclusively apply to discouragement now).

all that said, I don't want to hold up this PR. the benefit to this new functionality definitely outweighs a wonderful interface. my review comments suggest small changes in the language & documentation in attempt to make the functionality more apparent. if they seem reasonable, they can be incorporated here or can easily be done as followups. the only one I think should be done in this PR (if accepted), is renaming `CheckIfShouldDiscourage` to something that captures that the function initiates discouragement and disconnection.

curious to hear reactions. I'm happy to help move this PR forward however makes sense. Aside from☝️ I'm ready to leave a review ACK / If a rename made sense to people, I'd be happy to make the updates & offer the changes for you to pick up here / etc..

#### another review comment 
could we rename `getpeerinfo->banscore` field ([link](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/net.cpp#L115)) to `misbehaviorscore`? It seems to apply exclusively to discouragement. And/or update the description to something more helpful, like `The programmatically calculated score for protocol violations. Can lead to the peer being disconnected and/or marked as discouraged.`  

#### follow ups 
I agree with some of the review comments about possible followups and am collecting here for easy reference
* clear misbehaving filter without restarting node  (https://github.com/bitcoin/bitcoin/pull/19219#discussion_r439052719)
* consolidate `Ban()` and `IsBanned()` functions ([link](https://github.com/bitcoin/bitcoin/pull/19219/files#r442410269))
* remove `-banscore` config arg ([1](https://github.com/bitcoin/bitcoin/pull/19219/files#r443074487) [2](https://github.com/bitcoin/bitcoin/pull/19219#issuecomment-652684340))
* remove `getpeerinfo->banscore` field (this comment, [link](https://github.com/bitcoin/bitcoin/pull/19219#issuecomment-652684340))

sipa commented on src/banman.cpp:
```diff
@@ -68,43 +68,28 @@ void BanMan::ClearBanned()
 }
 
-bool BanMan::IsBanned(CNetAddr net_addr)
+bool BanMan::IsDiscouraged(const CNetAddr& net_addr)
```
Done.

sipa commented on src/net_processing.cpp:
```diff
@@ -3559,25 +3560,26 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
 }
 
-bool PeerLogicValidation::CheckIfBanned(CNode& pnode)
+bool PeerLogicValidation::CheckIfShouldDiscourage(CNode& pnode)
```
Changed to `MaybeDiscourageAndDisconnect`.

sipa commented on src/init.cpp:
```diff
@@ -427,8 +427,8 @@ void SetupServerArgs(NodeContext& node)
-    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-bantime=<n>", strprintf("Number of seconds to keep misbehaving peers from reconnecting (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting and discouraging misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-bantime=<n>", strprintf("Default bantime (in seconds) for manually configured bans (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Done.

sipa commented on doc/release-notes-19219.md:
```diff
@@ -0,0 +1,21 @@
+discouraged nodes in log output, as they're not (and weren't) strictly banned:
+incoming connections are still allowed from them, but they're preferred for
+eviction. Discouraged addresses are treated slightly different than banned
+ones:
```
Agree, the text didn't make that clear. I've changed it to "Furthermore, a few additional changes to how discouraged addressed are treated are introduced:". LMK what you think.

amitiuttarwar replied:
looks good

sipa commented on src/banman.h:
```diff
@@ -23,32 +24,55 @@ class CClientUIInterface;
-// dangerous, because it can cause a network split
-// between nodes running old code and nodes running
-// new code.
+// Banman manages two related but distinct concepts:
```
Thank @jnewbery!

sipa commented on src/net_processing.cpp:
```diff
@@ -1019,7 +1019,7 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans)
 
 /**
- * Mark a misbehaving peer to be banned depending upon the value of `-banscore`.
+ * Mark a misbehaving peer to be discouraged depending upon the value of `-banscore`.
```
Done.

sipa commented:
@amitiuttarwar Thanks for the review. I've addresses all of your comments, I think. As for your overall suggestion around "discouraging" as a term, do you mean that primarily in RPC/log interface, or in code comments? If there are small changes you see that can improve the interface it may be worth including them here, as they may become harder to change later.

Besides that, I'd like to get this merged.

sipa commented:
Oh, vaguely related: @gmaxwell and I working on designing/implementing a rolling cuckoo filter (see [this paper](https://www.cs.cmu.edu/~dga/papers/cuckoo-conext2014.pdf)), which would be a drop-in replacement for the rolling bloom filters we have, but around 3x smaller. However, they also support deletion, which may be useful if we'd ever want a set of RPCs to actually manage discouragement (separate from the ban management RPCs).

jnewbery commented:
> I don't feel like touching the banscore option and RPC field, as I think they should go away...

Totally agree on these. It seems to me that historically there are two reasons that configuration options get added:

1. Different users require different functionality from the system, and the configuration controls those different functionalities.
2. All (or _virtually all_) users require the same functionality, but the developer has punted (or hedged - pick how charitable you want to be) on the right parameters for the system.

`-banscore` certainly falls under the second category. As @sipa points out, we have no advice for users on when to change it, and it's not clear how you'd change it to achieve different desired outcomes.

We should aim to eliminate all of the second category, while being careful that there aren't genuine use cases for those config options.

amitiuttarwar commented:
> As for your overall suggestion around "discouraging" as a term, do you mean that primarily in RPC/log interface, or in code symbol names/comments? If there are small changes you see that can improve the interface it may be worth including them here, as they may become harder to change later.

Was referring to code internals too, so symbols/comments. I proposed my identified small changes to help with clarity (thanks for updates!) and agree moving forward with merge makes more sense than more rounds of changes and review. 

Also agree with reasoning & removing banscore config & RPC. Updated my comment aggregating follow ups to reflect. 

From my phone 😛 changes look good. I'm off the grid till Monday, but I'll re-review ack then if this is still open.

jnewbery commented on src/init.cpp:
```diff
@@ -430,8 +430,8 @@ void SetupServerArgs(NodeContext& node)
-    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-bantime=<n>", strprintf("Number of seconds to keep misbehaving peers from reconnecting (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-banscore=<n>", strprintf("Threshold for disconnecting and discouraging misbehaving peers (default: %u)", DEFAULT_BANSCORE_THRESHOLD), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-bantime=<n>", strprintf("Default time (in seconds) to disallow manually configured bans from reconnecting (default: %lu)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
why change this to %lu? `DEFAULT_MISBEHAVING_BANTIME` isn't a long

jonatack replied:
True, though the printed value is unchanged.

Proposed simplification if you retouch:

current
```
       Default time (in seconds) to disallow manually configured bans from
       reconnecting (default: 86400)
```
proposed
```
       Default duration (in seconds) of manually configured bans (default: 86400)
```

sipa replied:
Fixed. I suspect the %lu was just muscle memory...

jnewbery reviewed (commented):
Most of the recent changes look fine. One question about a string format specifier change.

jonatack reviewed (commented):
re-ACK 1f7e0ca per `git range-diff daae8b8 2605bdd 1f7e0ca` changes since last review: (a) edits to release note, bantime help and doxygen doc, (b) a function renaming, and (c) passing `CNetAddr net_addr` to `IsBanned()` by const reference instead of by value.

Would be good to merge this soon for 0.20 backport in #19224.

sipa commented on doc/release-notes-19219.md:
```diff
@@ -0,0 +1,23 @@
+incoming connections are still allowed from them, but they're preferred for
+eviction.
+
+Furthermore, a few additional changes to how discouraged addressed are
```
Done.

jonatack reviewed (commented):
ACK 2ad5838 per changes since last review `git range-diff 3276c14 1f7e0ca 2ad5838`

gmaxwell commented:
this fixes a publicly disclosed vulnerability and has been open for a month. Is it worth merging now?

jnewbery commented on src/net.cpp:
```diff
@@ -1010,17 +1010,24 @@ void CConnman::AcceptConnection(const ListenSocket& hListenSocket) {
-    // if the only banning reason was an automatic misbehavior ban.
-    if (!NetPermissions::HasFlag(permissionFlags, NetPermissionFlags::PF_NOBAN) && bannedlevel > ((nInbound + 1 < nMaxInbound) ? 1 : 0))
+    // Don't accept connections from banned peers.
+    bool banned = m_banman->IsBanned(addr);
```
This appears to be slightly less defensive since it's not checking that `m_banman` has been assigned before dereferencing it. It's fine because `m_banman` is always set in `CConnman::Init()` from the value in `connOptions.m_banman` (which is itself always set in `AppInitMain()`.

sipa replied:
Let's fix that later.

jnewbery reviewed (commented):
Code review ACK 2ad58381fffb33d611abf900b73d9e6b5a4e35f8

naumenkogs commented:
utACK 2ad58381fffb33d611abf900b73d9e6b5a4e35f8

amitiuttarwar commented:
code review ACK 2ad58381ff

fanquake commented:
I've added this for backport in #19224.
